### ISSUE TITLE
Error Moor on Profile mode but not in Debug mode

### ISSUE THREAD

**Issue by blackBoyCode** [simolus3/moor#690]
Hello,
I'm having an issue with a stream from my database. everything work fine in debug mode but in profile mode or release mode I get this error:
`"SqliteException(1555): UNIQUE constraint failed: bookmark_table.id, constraint failed (code 1555)"
E/flutter (17463): #0      PreparedStatement.execute (package:moor_ffi/src/impl/prepared_statement.dart:107)
E/flutter (17463): #1      _VmDelegate._runWithArgs (package:moor_ffi/src/vm_database.dart:72)
E/flutter (17463): #2      _VmDelegate.runInsert (package:moor_ffi/src/vm_database.dart:84)
E/flutter (17463): #3      _ExecutorWithQueryDelegate.runInsert.<anonymous closure> (package:moor/src/runtime/executor/helpers/engines.dart:69)
E/flutter (17463): #4      BasicLock.synchronized (package:synchronized/src/basic_lock.dart:32)
E/flutter (17463): #5      _ExecutorWithQueryDelegate._synchronized (package:moor/src/runtime/executor/helpers/engines.dart:22)
E/flutter (17463): #6      _ExecutorWithQueryDelegate.runInsert (package:moor/src/runtime/executor/helpers/engines.dart:67)
E/flutter (17463): #7      _MoorServer._runQuery (package:moor/src/runtime/isolate/server.dart:87)
E/flutter (17463): <asynchronous suspension>
E/flutter (17463): #8      IsolateCommunication.setRequestHandler.<anonymous closure>.<anonymous closure> (package:moor/src/runtime/isolate/communication.dart)
E/flutter (17463): <asynchronous suspension>`

This is the stream method I created inside the MyDatabase and currently using with a streambuilder in my UI :

**MyDatabase Stream Method**

``` 
Stream<bool> containsBookmarkStream(int id) {
    return select(
            bookmarkTable) 
        .watch()
        .map((event) => event.any((element) => element.id == id));
  }
```

**UI**
```
StreamBuilder<bool>(
          stream: database.containsBookmarkStream(currentInfoModelID),
          builder: (context, snapshot) {
            if (snapshot.hasData) {
              bool isSaved = snapshot.data;

              return IconButton(
                tooltip: 'Bookmark',
                icon: (isSaved)
                    ? Icon(
                        Icons.star,
                        color: iconColor,
                      )
                    : Icon(
                        Icons.star_border,
                        color: iconColor,
                      ),
                onPressed: () {
                  switch (!isSaved) {
                    case true:
                      database.addBookmark(currentInfoModelID);

                      break;
                    case false:
                      database.removeBookmark(currentInfoModelID);

                      break;
                  }

                  Scaffold.of(context).showSnackBar(
                    SnackBar(
                      duration: Duration(seconds: 1),
                      content: isSaved ? Text('Removed') : Text('Saved'),
                    ),
                  );
                },
              );
            } else {
              return CircularProgressIndicator();
            }
          },
        ),
```
notes: I'm using provider and also using isolate from this [example](https://moor.simonbinder.eu/docs/advanced-features/isolates/): 

Thanks for the help.

**Comment by simolus3** [comment]
Thanks for the report. Can you enable `logStatements` on the `VmDatabase` so that moor logs the statements it issues?

**Comment by blackBoyCode** [comment]
I 've enabled log statements for both debug mode and profile mode, here the result of the two:

**PROFILE MODE**

```
Moor: Sent BEGIN TRANSACTION with args []
I/flutter (20492): Moor: Sent INSERT INTO bookmark_table (id) VALUES (?) with args [100983271063552]
I/flutter (20492): Moor: Sent ROLLBACK TRANSACTION with args []
I/flutter (20492): Moor: Sent BEGIN TRANSACTION with args []
E/flutter (20492): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: "SqliteException(1555): UNIQUE constraint failed: bookmark_table.id, constraint failed (code 1555)"
E/flutter (20492): #0      PreparedStatement.execute (package:moor_ffi/src/impl/prepared_statement.dart:107)
E/flutter (20492): #1      _VmDelegate._runWithArgs (package:moor_ffi/src/vm_database.dart:72)
E/flutter (20492): #2      _VmDelegate.runInsert (package:moor_ffi/src/vm_database.dart:84)
E/flutter (20492): #3      _ExecutorWithQueryDelegate.runInsert.<anonymous closure> (package:moor/src/runtime/executor/helpers/engines.dart:69)
E/flutter (20492): #4      BasicLock.synchronized (package:synchronized/src/basic_lock.dart:32)
E/flutter (20492): #5      _ExecutorWithQueryDelegate._synchronized (package:moor/src/runtime/executor/helpers/engines.dart:22)
E/flutter (20492): #6      _ExecutorWithQueryDelegate.runInsert (package:moor/src/runtime/executor/helpers/engines.dart:67)
E/flutter (20492): #7      _MoorServer._runQuery (package:moor/src/runtime/isolate/server.dart:87)
E/flutter (20492): <asynchronous suspension>
E/flutter (20492): #8      IsolateCommunication.setRequestHandler.<anonymous closure>.<anonymous closure> (package:moor/src/runtime/isolate/communication.dart)
E/flutter (20492): <asynchronous suspension>
E/flutter (20492): 
..... 

```

**DEBUG MODE**

```
I/flutter (23767): Moor: Sent INSERT INTO bookmark_table (id) VALUES (?) with args [23515]
I/flutter (23767): Moor: Sent COMMIT TRANSACTION with args []
I/flutter (23767): Moor: Sent SELECT * FROM bookmark_table; with args []

.....


I/flutter (23767): Moor: Sent DELETE FROM bookmark_table WHERE id = ?; with args [23515]
I/flutter (23767): Moor: Sent COMMIT TRANSACTION with args []
I/flutter (23767): Moor: Sent SELECT * FROM bookmark_table; with args []

.....

```

It's seem to have a rollback transaction in Profile mode but not in Debug mode.

notes : the bookmark_table only contains integer ids inside with autoIncrement()

**Comment by simolus3** [comment]
This sounds a bit like https://github.com/flutter/flutter/issues/54948, are you building on Windows? And which Flutter version are you on?

When permuting the upper and lower 32 bits of `23515`, we get `100996155965440`. For `100983271063552` it would be `23512`.
Since that permutation shows up in the logs, it sounds like it happens in Dart, before calling ffi methods. If you `print(currentInfoModelID)`, does that show the same values on debug and profile builds?

**Comment by blackBoyCode** [comment]
I' ve run flutter doctor for the full details:


```
[√] Flutter (Channel stable, v1.17.5, on Microsoft Windows [Version 10.0.18363.900], locale en-US)
    • Flutter version 1.17.5 at C:\src\flutter
    • Framework revision 8af6b2f038 (11 days ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4


[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Users\Uns1t\AppData\Local\Android\sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[!] IntelliJ IDEA Ultimate Edition (version 2019.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2019.1.3
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[√] Connected device (1 available)
    • One • 57f688ec • android-arm • Android 6.0.1 (API 23)

! Doctor found issues in 1 category.

```

here is the print statements from both debug and profile mode:

**Profile mode**

```
....
I/flutter (21471): 100991860998144 // PRINTED VALUE HERE OF currentInfoModelID
I/flutter (21471): Moor: Sent BEGIN TRANSACTION with args []
I/flutter (21471): Moor: Sent INSERT INTO bookmark_table (id) VALUES (?) with args [100991860998144]
I/flutter (21471): Moor: Sent ROLLBACK TRANSACTION with args []
E/flutter (21471): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: "SqliteException(1555): UNIQUE constraint failed: bookmark_table.id, constraint failed (code 1555)"
E/flutter (21471): #0      PreparedStatement.execute (package:moor_ffi/src/impl/prepared_statement.dart:107)
E/flutter (21471): #1      _VmDelegate._runWithArgs (package:moor_ffi/src/vm_database.dart:72)
E/flutter (21471): #2      _VmDelegate.runInsert (package:moor_ffi/src/vm_database.dart:84)
E/flutter (21471): #3      _ExecutorWithQueryDelegate.runInsert.<anonymous closure> (package:moor/src/runtime/executor/helpers/engines.dart:69)
E/flutter (21471): #4      BasicLock.synchronized (package:synchronized/src/basic_lock.dart:32)
E/flutter (21471): #5      _ExecutorWithQueryDelegate._synchronized (package:moor/src/runtime/executor/helpers/engines.dart:22)
E/flutter (21471): #6      _ExecutorWithQueryDelegate.runInsert (package:moor/src/runtime/executor/helpers/engines.dart:67)
E/flutter (21471): #7      _MoorServer._runQuery (package:moor/src/runtime/isolate/server.dart:87)
E/flutter (21471): <asynchronous suspension>
E/flutter (21471): #8      IsolateCommunication.setRequestHandler.<anonymous closure>.<anonymous closure> (package:moor/src/runtime/isolate/communication.dart)
E/flutter (21471): <asynchronous suspension>
E/flutter (21471): ........ 


```

**Debug mode**

```
......
I/flutter (24192): 23514 //PRINTED VALUE HERE OF currentInfoModelID
I/flutter (24192): Moor: Sent BEGIN TRANSACTION with args []
I/flutter (24192): Moor: Sent INSERT INTO bookmark_table (id) VALUES (?) with args [23514]
I/flutter (24192): Moor: Sent COMMIT TRANSACTION with args []
I/flutter (24192): Moor: Sent SELECT * FROM bookmark_table; with args []
.......

```

you're right my id doesn't print 23514 in profile mode.

**Comment by simolus3** [comment]
I think that Flutter issue hasn't yet been fixed on `stable`. Can you try to switch to an unstable Flutter channel and see if that works? It switching to say `master` works, it's probably because of that compiler bug.

**Comment by blackBoyCode** [comment]
I've changed to `flutter channel master` but went I compile the code it's gave me a build error:

```
Launching lib\main.dart on One in release mode...
Running Gradle task 'assembleRelease'...
Error: Could not resolve the package 'characters' in 'package:characters/characters.dart'.
/C:/src/flutter/packages/flutter/lib/src/material/text_field.dart:9:8: Error: Not found: 'package:characters/characters.dart'
import 'package:characters/characters.dart';
       ^
/C:/src/flutter/packages/flutter/lib/widgets.dart:18:1: Error: Not found: 'package:characters/characters.dart'
export 'package:characters/characters.dart';
^
/C:/src/flutter/packages/flutter/lib/src/services/text_formatter.dart:9:8: Error: Not found: 'package:characters/characters.dart'
import 'package:characters/characters.dart';
       ^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:10:8: Error: Not found: 'package:characters/characters.dart'
import 'package:characters/characters.dart';
       ^
/C:/src/flutter/packages/flutter/lib/src/material/text_field.dart:822:61: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
  int get _currentLength => _effectiveController.value.text.characters.length;
                                                            ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/material/text_field.dart:824:118: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
  bool get _hasIntrinsicError => widget.maxLength != null && widget.maxLength > 0 && _effectiveController.value.text.characters.length > widget.maxLength;
                                                                                                                     ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/services/text_formatter.dart:355:11: Error: 'CharacterRange' isn't a type.
    final CharacterRange iterator = CharacterRange(value.text);
          ^^^^^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/services/text_formatter.dart:355:37: Error: Method not found: 'CharacterRange'.
    final CharacterRange iterator = CharacterRange(value.text);
                                    ^^^^^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/services/text_formatter.dart:356:20: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
    if (value.text.characters.length > maxLength) {
                   ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/services/text_formatter.dart:375:61: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
    if (maxLength != null && maxLength > 0 && newValue.text.characters.length > maxLength) {
                                                            ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/services/text_formatter.dart:378:25: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
      if (oldValue.text.characters.length == maxLength) {
                        ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:244:64: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
       assert(obscuringCharacter != null && obscuringCharacter.characters.length == 1),
                                                               ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:361:35: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
    assert(value != null && value.characters.length == 1);
                                  ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:567:11: Error: 'Characters' isn't a type.
    final Characters remaining = string.characters.skipWhile((String currentString) {
          ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:567:41: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
    final Characters remaining = string.characters.skipWhile((String currentString) {
                                        ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:600:47: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
    for (final String currentString in string.characters) {
                                              ^^^^^^^^^^
/C:/src/flutter/packages/flutter/lib/src/rendering/editable.dart:602:40: Error: The getter 'characters' isn't defined for the class 'String'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'characters'.
          !_isWhitespace(currentString.characters.first.toString().codeUnitAt(0))) {
                                       ^^^^^^^^^^
Unhandled exception:
FileSystemException(uri=org-dartlang-untranslatable-uri:package%3Acharacters%2Fcharacters.dart; message=StandardFileSystem only supports file:* and da
ta:* URIs)
#0      StandardFileSystem.entityForUri (package:front_end/src/api_prototype/standard_file_system.dart:33:7)
#1      asFileUri (package:vm/kernel_front_end.dart:659:37)
#2      writeDepfile (package:vm/kernel_front_end.dart:853:21)
<asynchronous suspension>
#3      FrontendCompiler.compile (package:frontend_server/frontend_server.dart:574:15)
<asynchronous suspension>
#4      _FlutterFrontendCompiler.compile (package:flutter_frontend_server/server.dart:43:22)
#5      starter (package:flutter_frontend_server/server
.dart:182:27)
#6      main (file:///C:/b/s/w/ir/cache/builder/src/flutter/flutter_frontend_server/bin/starter.dart:9:30)
#7      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
#8      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)



FAILURE: Build failed with an exception.

* Where:
Script 'C:\src\flutter\packages\flutter_tools\gradle\flutter.gradle' line: 896

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command 'C:\src\flutter\bin\flutter.bat'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2m 2s
Exception: Gradle task assembleRelease failed with exit code 1
```

I think I'll need to wait until it can be fixed in the stable channel . Meanwhile, I'll see if the're a workaround until it can be fixed but I wonder if I could change the moor_ffi to moor_flutter version? would it be the same result? is it recommended?

**Comment by simolus3** [comment]
You can try to use `moor_flutter` instead of `moor_ffi`, yes. `moor_ffi` is faster and allows us to use the latest sqlite3 version, but there's nothing wrong with using `moor_flutter`, especially if there are bugs preventing you from using `dart:ffi`.

**Comment by blackBoyCode** [comment]
This issue has been fixed in flutter 1.20 , so I'll close it. Thanks for the help, suggestions and the moor package since it's help me a lot during my development I appreciated it.
